### DISCUSSION TITLE
Training StyleGAN2 in Flax

### DISCUSSION THREAD

**Discussion by matthias-wright** [google/flax#1539] [Show and tell]
Hi all,

a while ago I posted here about some [pretrained models](https://github.com/google/flax/discussions/1320) that I ported to Flax. I have now added the training code for StyleGAN2: https://github.com/matthias-wright/flaxmodels/tree/main/training/stylegan2

**Highlights**
* Multi GPU training (data parallelism) with `jax.pmap`
* Results match the original implementation (see below)
* `tf.data.TFRecordDataset` is used for efficient data loading
* `optax` is used for optimization

**Lowlights**
* I only trained on resolution 512x512 due to limited resources
* Mixed-precision training is not stable yet. I used `flax.optim.DynamicScale` and also all the tricks from the original implementation (casting to float32 for some operations, using pre-normalization in the modulated conv layer, only using float16 for the higher resolutions, clipping the output of the convolution layers, etc) but it is not properly working yet.

I also added instructions on how to train on your own dataset. If you have any questions feel free to ask them here or on the [github](https://github.com/matthias-wright/flaxmodels) page.

Cheers!

P.S. I also added training code for [ResNet](https://github.com/matthias-wright/flaxmodels/tree/main/training/resnet) and [VGG](https://github.com/matthias-wright/flaxmodels/tree/main/training/vgg) but that is less interesting since there are already some great examples [here](https://github.com/google/flax/tree/main/examples/imagenet), which I also borrowed from. 

**Some results**
![alt text](https://raw.githubusercontent.com/matthias-wright/flaxmodels/main/training/stylegan2/images/ffhq_overview.jpg)

**Style mixing**
![alt text](https://raw.githubusercontent.com/matthias-wright/flaxmodels/main/training/stylegan2/images/ffhq_style_mixing.jpg)
